src/tribler/core/utilities/search_utils.py

```diff
@@ -27,3 +32,194 @@ def split_into_keywords(string, to_filter_stopwords=False):
 
 def filter_keywords(keywords):
     return [kw for kw in keywords if len(kw) > 0 and kw not in DIALOG_STOPWORDS]
+
+
+def item_rank(query: str, item: dict):
+    title = item['name']
+    seeders = item.get('num_seeders', 0)
+    leechers = item.get('num_leechers', 0)
+    freshness = time.time() - item.get('updated', 0)
+    return torrent_rank(query, title, seeders + leechers * 0.1, freshness)
+
+
+def torrent_rank(query: str, title: str, seeders: int = 0, freshness: Optional[float] = 0) -> float:
+    """
+    Calculates search rank for a torrent.
+    Takes into account:
+      - similarity of the title to the query string;
+      - the reported number of seeders;
+      - how long ago the torrent file was created.
+    """
+    freshness = max(0, freshness or 0)
+    tr = title_rank(query or '', title or '')
+    sr = (seeders_rank(seeders or 0) + 9) / 10  # range [0.9, 1]
+    fr = (freshness_rank(freshness) + 9) / 10  # range [0.9, 1]
+    result = tr * sr * fr
+    # uncomment the next line to debug the function inside an SQL query:
+    # print(f'*** {result} : {seeders}/{freshness} ({freshness / SECONDS_IN_DAY} days)/{title} | {query}')
+    return result
+
+
+def seeders_rank(seeders: float) -> float:
+    """
+    Calculates rank based on the number of seeders. The result is normalized to the range [0, 1]
+    """
+    return seeders / (100 + seeders)  # inf seeders -> 1; 100 seeders -> 0.5; 10 seeders -> approx 0.1
+
+
+def freshness_rank(freshness: Optional[float] = 0):
+    """
+    Calculates rank based on the torrent freshness. The result is normalized to the range [0, 1]
```

Comment:
I added a description to the code. Freshness is the number of seconds since the torrent creation. It is more convenient to use than a timestamp, as, with freshness, it is not necessary to calculate the current time inside the function, and tests are a bit easier.

I have seen a few torrents in the database whose timestamps are in the future, and their calculated freshness is negative. If a freshness value is less or equal to zero, it is considered to be invalid, and the resulting rank is zero.

Reply:
🙏